extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/connection/ConnectionService.java

```diff
@@ -113,56 +123,65 @@
-            while (results.hasMore()) {
-
-                try {
-
-                    LDAPEntry entry = results.next();
-
-                    // Get common name (CN)
-                    LDAPAttribute cn = entry.getAttribute("cn");
-                    if (cn == null) {
-                        logger.warn("guacConfigGroup is missing a cn.");
-                        continue;
-                    }
+            while (results.next()) {
+
+                // Get the entry
+                Response response = results.get();
+                Entry entry;
+                if (response instanceof SearchResultEntry)
+                    entry = ((SearchResultEntry)results).getEntry();
+                else
+                    continue;
+
+                // Get common name (CN)
+                Attribute cn = entry.get("cn");
+                if (cn == null) {
+                    logger.warn("guacConfigGroup is missing a cn.");
+                    continue;
+                }
 
-                    // Get associated protocol
-                    LDAPAttribute protocol = entry.getAttribute("guacConfigProtocol");
-                    if (protocol == null) {
-                        logger.warn("guacConfigGroup \"{}\" is missing the "
-                                  + "required \"guacConfigProtocol\" attribute.",
-                                cn.getStringValue());
-                        continue;
-                    }
+                // Get associated protocol
+                Attribute protocol = entry.get("guacConfigProtocol");
+                if (protocol == null) {
+                    logger.warn("guacConfigGroup \"{}\" is missing the "
+                              + "required \"guacConfigProtocol\" attribute.",
+                            cn.getString());
+                    continue;
+                }
 
-                    // Set protocol
-                    GuacamoleConfiguration config = new GuacamoleConfiguration();
-                    config.setProtocol(protocol.getStringValue());
+                // Set protocol
+                GuacamoleConfiguration config = new GuacamoleConfiguration();
+                config.setProtocol(protocol.getString());
 
-                    // Get parameters, if any
-                    LDAPAttribute parameterAttribute = entry.getAttribute("guacConfigParameter");
-                    if (parameterAttribute != null) {
+                // Get parameters, if any
+                Attribute parameterAttribute = entry.get("guacConfigParameter");
+                if (parameterAttribute != null) {
 
-                        // For each parameter
-                        Enumeration<?> parameters = parameterAttribute.getStringValues();
-                        while (parameters.hasMoreElements()) {
+                    // For each parameter
+                    Iterator parameters = parameterAttribute.iterator();
```

Comment:
Why not `Iterator<String>`? Or does the new API not use generics?

Reply:
Yeah, I think the API does not implement any method that allows for retrieving all of the string values at one time, so instead you have to get the iterator and got through and retrieve each value.  Let me know if you see something that makes you think there's an alternate way to do this.

Reply:
I don't see any problem with using an iterator - the issue here is that the type parameter for `Iterator` is not being used, and thus we are later having to typecast to `String`. Checking the API for [`Attribute`](http://directory.apache.org/api/gen-docs/latest/apidocs/org/apache/directory/api/ldap/model/entry/Attribute.html), the `iterator()` function returns an `Iterator<Value<?>>`, with `Value<?>` providing a `getString()`. I'm not sure why typecasting to `String` would be working, but I think we should be using proper Java generics.

Reply:
By the way, for anything implementing `Iterable`, you can use a special `for` syntax to iterate through its contents. In this case, with `Iterable<Value<?>>`, you can:

    for (Value<?> value : parameterAttribute) {
        ...
    }

There's generally no need to obtain the iterator itself unless you're going to be altering the contents of the iterated object through invoking `iterator.remove()`.

Reply:
Fixed, x2.